### PR TITLE
Prepared Library For NPM Publication

### PR DESCRIPTION
**Pull request by kelvinator07**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,36 @@
...
+### Install dependencies
+
+> npm install
+
+## Usage
+
+To build the tree, simply call the `buildPMT` function and pass the `leaves` as a list of strings and a `filteredHash` in string format, i.e buildPMT(leaves: string[], filteredHash: string)
+
+First parameter {leaves}: An array of transaction hash of all transactions in a block. 
+Second parameter {filteredHash}: transaction hash (the hash without witness) of the coinbase transaction.
```
Comment (marcos-iov) [inline README.md:22]: It doesn't have to be the coinbase transaction. The PMT can be built for any transaction

File: README.md [hunk 2]
```diff
@@ -1,2 +1,36 @@
...
+
+This library can be used to build a partial merkle tree for a block of transactions. A partial merkle tree is a pruned version of a full merkle tree, containing only the parts needed to prove a given transaction is part of the tree. 
+
+## Installation
+
+> npm install @rsksmart/pmt-builder
+
+### Install dependencies
+
+> npm install
```
Comment (marcos-iov) [inline README.md:15]: What's the difference between this and the above?

**Review by jeremy-then — APPROVED** [review]
LGTM.

File: README.md [hunk 3]
```diff
@@ -1,2 +1,36 @@
...
+
+To build the tree, simply call the `buildPMT` function and pass the `leaves` as a list of strings and a `filteredHash` in string format, i.e buildPMT(leaves: string[], filteredHash: string)
+
+First parameter {leaves}: An array of transaction hash of all transactions in a block. 
+Second parameter {filteredHash}: transaction hash (the hash without witness) of the coinbase transaction.
+
+The function returns an object with the following fields:
+totalTX: the size of {leaves} in the array.
+Hashes: All the hashes that build the partial merkle tree. This includes the leaves level as well as the upper levels of the tree.
+flags: ???
```
Comment (marcos-iov) [inline README.md:27]: I don't remember what this meant, to be honest. If you can't find any info online just leave it blank

File: sample/sample.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    '6cc629897f3f6c3e873e98af8061f54618fe2be6599853841ca97aeb9614a136',
+    '31c8227345ead8477845fda6d8dc68cc1bca9c094523aef60d857d521aca6937',
+    'e0714cbc5ed04014e1a1f362612fdb05578a1c8e45257ac6cc0dff0d2e532855',
+    '0f9e8932b5bbcad80242077205843e42bcca9d84c58b751535ea07a06e4179e6',
+    '5a4d2446ff81be95eacaaac0fcd06856f9625f6f07391a2bf90b1a065be0e0fc'
+];
+
+const resultPmt = pmtBuilder.buildPMT(blockTransactions, blockTransactions[0]);
+
+console.log(resultPmt);
```
Comment (marcos-iov) [inline sample/sample.js:14]: Maybe print the block transactions and the filtered hash in console before the result. So users can understand where the pmt is coming from

File: README.md [hunk 5]
```diff
@@ -1,2 +1,36 @@
...
+Hex: The partial merkle tree serialized in hex format.
+
+## Test
+
+To run test:
+
+> npm test
+
```
Comment (marcos-iov) [inline README.md:35]: Add a sample snippet maybe?

```
const pmtBuilder = require("@rsksmart/pmt-builder");

const blockTransactions =[
'8f01832aa125683490e70a9142ccd9c49485b84708180487b5f35dc7795a3afd',
'6cc629897f3f6c3e873e98af8061f54618fe2be6599853841ca97aeb9614a136',
'31c8227345ead8477845fda6d8dc68cc1bca9c094523aef60d857d521aca6937',
'e0714cbc5ed04014e1a1f362612fdb05578a1c8e45257ac6cc0dff0d2e532855',
'0f9e8932b5bbcad80242077205843e42bcca9d84c58b751535ea07a06e4179e6',
'5a4d2446ff81be95eacaaac0fcd06856f9625f6f07391a2bf90b1a065be0e0fc'
];

const resultPmt = pmtBuilder.buildPMT(blockTransactions, blockTransactions[0]);
```

File: README.md (hunk 1 above)
Comment (kelvinator07) [reply README.md]: Resolved

File: README.md (hunk 2 above)
Comment (kelvinator07) [reply README.md:15]: This Installs The node_modules

File: sample/sample.js (hunk 4 above)
Comment (kelvinator07) [reply sample/sample.js]: Resolved

File: README.md (hunk 5 above)
Comment (kelvinator07) [reply README.md:53]: Resolved

File: README.md (hunk 3 above)
Comment (kelvinator07) [reply README.md]: Updated

File: README.md [hunk 6]
```diff
@@ -1,2 +1,54 @@
...
+### Install dependencies
+
+> npm install
+
+## Usage
+
+To build the tree, simply call the `buildPMT` function and pass the `leaves` as a list of strings and a `filteredHash` in string format, i.e buildPMT(leaves: string[], filteredHash: string)
+
+First parameter {leaves}: An array of transaction hash of all transactions in a block.
```
Comment (nathanieliov) [inline README.md:21]:
```suggestion
First parameter {leaves}: An array of transaction hash without witness(txid) of all transactions in a block. 
```

File: README.md [hunk 7]
```diff
@@ -1,2 +1,54 @@
...
+
+To build the tree, simply call the `buildPMT` function and pass the `leaves` as a list of strings and a `filteredHash` in string format, i.e buildPMT(leaves: string[], filteredHash: string)
+
+First parameter {leaves}: An array of transaction hash of all transactions in a block. 
+Second parameter {filteredHash}: transaction hash (the hash without witness).
+
+The function returns an object with the following fields:
+totalTX: the size of {leaves} in the array.
+Hashes: All the hashes that build the partial merkle tree. This includes the leaves level as well as the upper levels of the tree.
+flags: The number of flags bytes to follow.
```
Comment (nathanieliov) [inline README.md:27]: What do you mean by "flags bytes to follow"? Honestly, I didn't get your point.

File: README.md [hunk 8]
```diff
@@ -1,2 +1,54 @@
...
+    '8f01832aa125683490e70a9142ccd9c49485b84708180487b5f35dc7795a3afd',
+    '6cc629897f3f6c3e873e98af8061f54618fe2be6599853841ca97aeb9614a136',
+    '31c8227345ead8477845fda6d8dc68cc1bca9c094523aef60d857d521aca6937',
+    'e0714cbc5ed04014e1a1f362612fdb05578a1c8e45257ac6cc0dff0d2e532855',
+    '0f9e8932b5bbcad80242077205843e42bcca9d84c58b751535ea07a06e4179e6',
+    '5a4d2446ff81be95eacaaac0fcd06856f9625f6f07391a2bf90b1a065be0e0fc'
+];
+
+const resultPmt = pmtBuilder.buildPMT(blockTransactions, blockTransactions[0]);
+console.log("Result: ", resultPmt);
```
Comment (nathanieliov) [inline README.md:45]:
```suggestion
console.log("Result: ", JSON.stringify(resultPmt, null, 2));
```

File: index.d.ts [hunk 9]
```diff
@@ -0,0 +1,6 @@
+/**
+  * 
+  * @param leaves array of transaction hashes in a block
```
Comment (nathanieliov) [inline index.d.ts:3]:
```suggestion
  * @param leaves array of transaction hashes without witness(txid) in a block
```

**Review by nathanieliov — COMMENTED** [review]
Well done K! Just left some minor comments.

File: README.md (hunk 7 above)
Comment (kelvinator07) [reply README.md:27]: See Reference: https://reference.cash/protocol/network/messages/merkleblock

File: index.d.ts (hunk 9 above)
Comment (kelvinator07) [reply index.d.ts]: Resolved

File: README.md (hunk 6 above)
Comment (kelvinator07) [reply README.md]: Resolved

File: README.md (hunk 8 above)
Comment (kelvinator07) [reply README.md:45]: The Result Isn't Deeply Nested, So JSON.stringify() Produces The Same Result With or Without

**Review by nathanieliov — APPROVED** [review]

File: README.md (hunk 7 above)
Comment (marcos-iov) [reply README.md:27]: @kelvinator07 in this case our `flags` field represents the `flag bits` described in the link you shared. Not the flag field length

**Review by marcos-iov — APPROVED** [review]

**Pull request merged**